x/incentives/keeper/gauge.go

```diff
@@ -110,13 +111,41 @@ func (k Keeper) CreateGauge(ctx sdk.Context, isPerpetual bool, owner sdk.AccAddr
 		}
 	}
 
-	// check if denom this gauge pays out to exists on-chain
-	if !k.bk.HasSupply(ctx, distrTo.Denom) && !strings.Contains(distrTo.Denom, "osmovaloper") {
-		return 0, fmt.Errorf("denom does not exist: %s", distrTo.Denom)
+	nextGaugeId := k.GetLastGaugeID(ctx) + 1
+
+	// For no lock gauges, a pool id must be set.
+	// A pool with such id must exist and be a concentrated pool.
+	if distrTo.LockQueryType == lockuptypes.NoLock {
+		if poolId == 0 {
+			return 0, fmt.Errorf("no lock gauges must have a pool id")
+		}
+
+		pool, err := k.pmk.GetPool(ctx, poolId)
+		if err != nil {
+			return 0, err
+		}
+
+		if pool.GetType() != poolmanagertypes.Concentrated {
+			return 0, fmt.Errorf("no lock gauges must be created for concentrated pools only")
+		}
+
+		// We assume that external gauges are created with 0 duration,
+		// while internal gauges are created with an incentive epoch duration.
+		k.pik.SetPoolGaugeId(ctx, poolId, distrTo.Duration, nextGaugeId)
+	} else {
+		// For all other gauges, pool id must be 0.
+		if poolId != 0 {
+			return 0, fmt.Errorf("pool id must be 0 for gauges with lock")
+		}
+
+		// check if denom this gauge pays out to exists on-chain
+		if !k.bk.HasSupply(ctx, distrTo.Denom) && !strings.Contains(distrTo.Denom, "osmovaloper") {
```

Comment:
I don't think we need to require specifying this denom for the "NoLock" option because it has no use. That's why I decided to move it inside the "Lock" branch only

Reply:
that makes sense so  this shouldn't be necessary 
https://github.com/osmosis-labs/osmosis/pull/5459/files#diff-968d496699ec6c793045c4d26a3a3de4af2f46e5a5d9a94b81ae11e0e59a1903R113

Reply:
From latest changes, I made the requirement for the denom to be unset for "NoLock" gauges so that the system overwrites it wit  the following denom `types.NoLockExternalGaugeDenom(poolId)`

This is useful as, internally, a store mapping will be created, utilising this denom and allowing us to query external gauges by pool id and this denom format

Reply:
I know this is a copy over from existing logic, but why would the distrTo.Denom contain osmovaloper?

Reply:
@czarcas7ic you asked the same question a while ago and i think you came to this conclusion
https://osmosis-network.slack.com/archives/C027ELA290B/p1682711900968259

Reply:
I think the purpose of this is to allow synth denoms that don't have supplies that exist on chain to be paid out to, but would be nice to have a comment to confirm this

Reply:
I have confirmed that this is for the synth denoms by setting a breakpoing here and running a superfluid test suite.

The denom I observed was:
```
"cl/pool/1/superbonding/osmovaloper1wcfyglfgjs2xtsyqu7pl60d0mpw5g7f4wh7pnm"
```

Reply:
[d3ce09a](https://github.com/osmosis-labs/osmosis/pull/5459/commits/d3ce09ab4ef16b13e5d18a4c14abe4260e6f7926)